design-decisions/2018-10-12-Condition-Split.md

```diff
@@ -0,0 +1,96 @@
+Splitting configurations upon new constraints on symbolic variables
+===================================================================
+
+Background
+----------
+
+When applying an axiom on a symbolic configuration new constraints may be
+introduced on symbolic variables, coming from side conditions, or directly
+from the unification process.
+
+### Example (from side conditions)
+
+Consider rewriting term `3 / A:Int` using
+
+`rule X:Int / Y:Int => X /Int Y requires Y =/=Int 0`
+which would be written as
+`Y =/= 0 = true /\ X:Int / Y:Int => X /Int Y`
+
+Then `3 / A:Int /\ (Y =/=Int 0 = true /\ X:Int / Y:Int)` will simplify to
+```
+3 / A:Int /\ (X:Int = 3 /\ Y:Int = A:Int)
+          /\ (A:Int =/=Int 0) = true
+```
+which is satisfiable, but not `T`.
+
+### Example (from unification)
+
+Consider rewriting term `if A:Int <Int 0 then -1 else 1` using
+
+`rule if true then E1:Exp else E2:Exp => E1`
+
+Then `if A:Int <Int 0 then -1 else 1 /\ if true then E1:Exp else E2:Exp` will
+simplify to
+```
+if A:Int <Int 0 then -1 else 1 /\ (E1:Exp = -1 /\ E2:Exp = 1)
+                               /\ A:Int <Int 0 = true
+```
+which is satisfiable, but not `T`.
+
+Problem
+-------
+
+Simply continuing the rewriting process when a new constraint is introduced
+(on some symbolic variable) is not sound when the constraint does not simplify
+to top.  The reason is that, going further, the coverage is reduced to the
+inputs which satisfy the constraint.
+
+Questions
+---------
+
+* When/how do we decide that a new constraint is satisfiable but not top?
+
+* How about such constraints coming from function application?
+
+Decision: do not apply function axioms if they introduce new constraints
+------------------------------------------------------------------------
+
+We might revert this decision later, but for now this is a safe bet, as it
+prevents undesirable unrolling of functions.
```

Comment:
Here, what does it even mean to "apply a function" if it introduces new constraints? Soundness shouldn't be at risk here. If it's just a policy for whether or not we try to unfold function applications it might be a reasonable policy. Are you making a distinction between "new constraints" added as a side condition and the constraints arising from unifying a non-trivial LHS with the configuration?
I think you probably mean to allow function applications if unification reports a successful pattern matching. But, if that's the intent you should write that explicitly - from a logical point of view I don't see any fundamental distinction between a pattern and a side condition, and regard unification success as just a special case of a rule stated as "only apply function axioms if new constraints are implied by existing ones".

Regarding soundness, given a current configuration `C[... f(X) ...] /\ P` applying a rule for `f` would be witnessed by a proof of something like `C[... f(X) ...]  /\ P = C[... rhs of rule ...] /\ P` saying the new expression for the configuration after unrolling the function is equal to the old. If the rule for `f` was instead `side-condition(X) /\ (f(X) = rhs of rule)` then we could only prove the bare equation `C[... f(X) ...] /\ P = C[... rhs of rule ...] /\ P` if we can prove that `P` implies `side-condition(X)`, otherwise we could only prove an equation like `C[... f(X) ...] /\ P = ((C[... rhs of rule ...] /\ P /\ side-condition(X)) \/ (C[... f(X) ...] /\ P /\ ~side-condition(X))`